File: trunk/src/wcsp.cpp
/*
 * Copyright 2008 Luděk Cigler <luc@matfyz.cz>
 * $Id$
 *
 * This file is part of SCSPSampler.
 *
 * SCSPSampler is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Hollo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <deque>
#include <iostream>
#include <fstream>
#include <sstream>
#include "math.h"

#include "csp.h"
#include "wcsp.h"

const double EXP_ROOT = 2.0;
double EXP_K = 0.001;

double WCSPConstraint::operator()(Assignment &a) const {
        std::vector<VarType> scopeAssignment;
        for (Scope::iterator scIt = mScope.begin(); scIt != mScope.end(); ++scIt) {
               scopeAssignment.push_back(a[*scIt]); 
        }

        unsigned int weight;
        std::map<std::vector<VarType>, unsigned long>::const_iterator weightIt = mDifferentWeightTuples.find(scopeAssignment);

        if (weightIt != mDifferentWeightTuples.end()) {
                weight = weightIt->second;
        } else {
                weight = mDefaultWeight;
        }

        if (weight >= mHardConstraintWeight) {
                return 0;
        } else {
                return exp(log(EXP_ROOT) * EXP_K * (-(double)weight));
        }
}

Scope WCSPConstraint::getScope() const {
        return mScope;
}

bool WCSPConstraint::isSoft() const {
        return mMaxTupleWeight < mHardConstraintWeight;
}

bool WCSPConstraint::hasSupport(VarIdType aVarId, VarType aValue, const CSPProblem &aProblem, const Assignment &aEvidence) {
        if (isSoft()) {
                return true;
        }

        Assignment::const_iterator aIt = aEvidence.find(aVarId);
        if (aIt != aEvidence.end()) {
                return aIt->second == aValue;
        }

        Assignment a = aEvidence;
        a[aVarId] = aValue;
        
        std::deque<VarIdType> variablesToAssign;

        for (Scope::iterator scIt = mScope.begin(); scIt != mScope.end(); ++scIt) {
                if (a.find(*scIt) == a.end())
                        variablesToAssign.push_back(*scIt);
        }

        return _hasSupportInternal(a, aProblem, variablesToAssign);
}


bool WCSPConstraint::_hasSupportInternal(Assignment &aEvidence, const CSPProblem &aProblem, std::deque<VarIdType> & aVarsToAssign) {
        if (aVarsToAssign.empty()) {
                std::vector<VarType> varAssignment;

                for (Scope::iterator scIt = mScope.begin(); scIt != mScope.end(); ++scIt) {
                        varAssignment.push_back(aEvidence[*scIt]);
                }

                // Now an assignment is allowed if it is not in mDisallowedTuples and the default value is not hard-constraint
                if (mDisallowedTuples.find(varAssignment) == mDisallowedTuples.end()) {
                        if (mDifferentWeightTuples.find(varAssignment) == mDifferentWeightTuples.end()) {
                                return (mDefaultWeight < mHardConstraintWeight);
                        } else {
                                return true;
                        }
                } else {
                        return false;
                }
        } else {
                VarIdType varId = aVarsToAssign.front();
                aVarsToAssign.pop_front();
                const Domain * d = aProblem.getVariableById(varId)->getDomain();

                for (Domain::const_iterator domIt = d->begin(); domIt != d->end(); ++domIt) {
                        aEvidence[varId] = *domIt;

                        if (_hasSupportInternal(aEvidence, aProblem, aVarsToAssign))
                                return true; // Once we find support, we don't need to search any further

                        aEvidence.erase(varId);
                }

                aVarsToAssign.push_front(varId);
                
                return false;
        }
}

CSPProblem * load_wcsp_problem(const char * fileName) {
        std::ifstream file(fileName);
        std::string line;

        unsigned int totalVariables, maxDomainSize, totalConstraints;
        unsigned long hardConstraintWeight;

        if (std::getline(file, line)) {
                std::vector<std::string> words;
                tokenize(line, words);

                totalVariables = parseArg<unsigned int>(words[1]);
                maxDomainSize = parseArg<unsigned int>(words[2]);
                totalConstraints = parseArg<unsigned int>(words[3]);
                hardConstraintWeight = parseArg<unsigned long>(words[4]);
        } else {
                return NULL;
        }

        VariableMap * variables = new VariableMap();

        if (std::getline(file, line)) {
                std::istringstream lineStream(line);
                unsigned long domainSize;
                VarIdType varId = 0;

                while (lineStream >> domainSize) {
                        // Create a new variable with domain of <0, domainSize - 1>
                        (*variables)[varId] = new Variable(varId, 0, domainSize - 1);
                        ++varId;
                }
        } else {
                return NULL;
        }


        ConstraintList * constraints = new ConstraintList();
        while (std::getline(file, line)) {
                // Read constraint specification line
                std::vector<std::string> words;
                tokenize(line, words);

                unsigned int constraintArity = parseArg<unsigned int>(words[0]);
                Scope scope;

                for (unsigned int scopeIdx = 0; scopeIdx < constraintArity && scopeIdx < words.size() - 2; ++scopeIdx) {
                        scope.insert(parseArg<VarIdType>(words[scopeIdx + 1]));
                }
                unsigned int defaultWeight = parseArg<unsigned int>(words[words.size() - 2]);
                unsigned int numDifferentTuples = parseArg<unsigned int>(words[words.size() - 1]);

                WCSPConstraint * ctr = new WCSPConstraint(scope, defaultWeight, hardConstraintWeight);

                for (unsigned int i = 0; i < numDifferentTuples; ++i) {
                        if (std::getline(file, line)) {
                                std::vector<VarType> tuple; 
                                std::vector<std::string> tupleWords;
                                tokenize(line, tupleWords);

                                for (unsigned j = 0; j < constraintArity; ++j) {
                                       tuple.push_back(parseArg<VarType>(tupleWords[j]));
                                }
                                unsigned int tupleWeight = parseArg<unsigned int>(tupleWords[tuple.size() - 1]);

                                ctr->addDifferentWeightTuple(tuple, tupleWeight);
                        } else {
                                break;
                        }
                }

                constraints->push_back(ctr);
        }

        return new CSPProblem(variables, constraints);
}

File: trunk/py/wcsp.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
##
## Copyright (C) 2008 Luděk Cigler <luc@matfyz.cz>
##
## This file is part of SCSPSampler.
##
## SCSPSampler is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or
## (at your option) any later version.
##
## Hollo is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program. If not, see <http://www.gnu.org/licenses/>.
##

import re
import os.path
from optparse import OptionParser
import random
import math

import psyco
psyco.full()


WCSP_PROBLEM_LINE = re.compile('^(?P<name>\S+)\s+(?P<num_vars>\d+)\s+(?P<max_domain_size>\d+)\s+(?P<total_constraints>\d+)\s+(?P<upper_bound>\d+)')

EXP_ROOT = 2.0
EXP_K = 0.001

class WCSPVariable:
    def __init__(self, var_id, domain_size):
        self._var_id = var_id
        self._domain = range(0, domain_size)

    def random_value(self):
        return random.choice(self._domain)

class WCSPConstraint:
    def __init__(self, scope, default_cost, upper_bound):
        self._scope = scope
        self._default_cost = int(default_cost)
        self._upper_bound = upper_bound
        self._disallowed_tuples = set()
        self._different_tuples = {}

    def add_different_tuple(self, a_tuple, cost):
        self._different_tuples[a_tuple] = cost
        if cost >= self._upper_bound:
            # Add disallowed tuple
            self._disallowed_tuples.add(a_tuple)

    def eval_assignment(self, assignment):
        a_tuple = []
        for val in self._scope:
            a_tuple.append(assignment[val])

        a_tuple = tuple(a_tuple) 
        if a_tuple in self._disallowed_tuples:
            return 0
        else:
            cost = (self._different_tuples.has_key(a_tuple) and self._different_tuples[a_tuple]) or self._default_cost
            return math.exp(math.log(EXP_ROOT) * EXP_K * (-cost))

class WCSPProblem:
    def __init__(self, filename):
        self._variables = []
        self._constraints = []
        file = open(filename)

        line = file.readline()
        # Read the WCSP prologue
        m = WCSP_PROBLEM_LINE.match(line)
        if not m:
            return
        m_groups = m.groupdict()
        self._num_vars = int(m_groups['num_vars'])
        self._max_domain_size = int(m_groups['max_domain_size'])
        self._total_constraints = int(m_groups['total_constraints'])
        self._upper_bound = int(m_groups['upper_bound'])

        # Load variables
        line = file.readline()
        domain_sizes = line.split()
        self._variables = [WCSPVariable(var_id, int(domain_size)) for domain_size, var_id in zip(domain_sizes, xrange(0, len(domain_sizes)))]

        while file:
            ctr_line = file.readline()
            ctr_data = ctr_line.split()
            if len(ctr_data) <= 0:
                break

            default_cost = ctr_data[-2]
            num_different_tuples = int(ctr_data[-1])
            c = WCSPConstraint(set([int(x) for x in ctr_data[1:-2]]), default_cost, self._upper_bound)
            for i in xrange(0, num_different_tuples):
                tuple_line = file.readline()
                tuple_line = tuple_line.split()
                tuple_cost = int(tuple_line[-1])
                tuple_data = [int(x) for x in tuple_line[0:-1]]
                c.add_different_tuple(tuple(tuple_data), tuple_cost)
            self._constraints.append(c)

    def eval_assignment(self, assignment):
        result = 1.0
        for c in self._constraints:
            result = result * c.eval_assignment(assignment)
        return result

    def random_sample(self, MAX_SAMPLES):
        for i in xrange(0, MAX_SAMPLES):
            sample = []
            for var in self._variables:
                sample.append(var.random_value())
            yield sample
                
if __name__ == '__main__':
    usage = "%prog [OPTIONS]"
    parser = OptionParser(usage=usage, version= "%prog 0.1")
    parser.add_option('-d', '--dataset',
                      action='store',
                      dest='dataset',
                      type='string',
                      metavar='PATH',
                      default='/home/luigi/Matfyz/Diplomka/scspsampler/trunk/data/wcsp/celar_small.wcsp',
                      help="PATH to data set")

    parser.add_option('-n', '--numsamples',
                      action='store',
                      dest='numsamples',
                      type='int',
                      metavar='PATH',
                      default='1000',
                      help="Number of generated samples")

    parser.add_option('-k', '--koef',
                      action='store',
                      dest='koef',
                      type='float',
                      metavar='VALUE',
                      default='0.01',
                      help="Coefitient of the algorithm")

    (options, args) = parser.parse_args()

    problem = WCSPProblem(options.dataset)
    EXP_K = options.koef

    for sample in problem.random_sample(options.numsamples):
        e = problem.eval_assignment(sample)
        if e > 0.0:
            print "%e, %s" % (e, ", ".join([str(x) for x in sample]))

